File: webapp/src/export/tripExportCodec.ts
import type { Trip, TripDay, ChecklistItem, ItemKind, Place } from "../model/trip";

// ── date helpers ───────────────────────────────────────────────
// trip/day dates: epoch-seconds ↔ "YYYY-MM-DD" (UTC date-only).
const epochToDateOnly = (sec: number): string => new Date(sec * 1000).toISOString().slice(0, 10);
const dateOnlyToEpoch = (s: string): number => Math.floor(Date.parse(`${s}T00:00:00Z`) / 1000);
// reminderDate: epoch-seconds ↔ ISO-8601 WITHOUT millis (matches Swift ISO8601DateFormatter).
const epochToIso = (sec: number): string => new Date(sec * 1000).toISOString().replace(/\.\d{3}Z$/, "Z");
const isoToEpoch = (s: string): number => Math.floor(Date.parse(s) / 1000);

// ── wire DTOs (canonical JSON shape) ───────────────────────────
interface PlaceDTO { name: string; query: string; latitude: number | null; longitude: number | null; }
interface DayDTO { date: string; city: string; title: string; }
interface ItemDTO {
  kind: string; label: string; notes: string; dayIndex: number | null; time: string | null;
  owner: string | null; isDone: boolean; sortOrder: number; reminderDate: string | null; place: PlaceDTO | null;
}
interface TripDTO {
  schemaVersion: number; name: string; startDate: string; endDate: string;
  destinations: string[]; days: DayDTO[]; items: ItemDTO[];
}

// ── JSON ───────────────────────────────────────────────────────
export function exportJSON(trip: Trip): string {
  const dayIndex = new Map(trip.days.map((d, i) => [d.id, i]));
  const dto: TripDTO = {
    schemaVersion: 1,
    name: trip.name,
    startDate: epochToDateOnly(trip.startDate),
    endDate: epochToDateOnly(trip.endDate),
    destinations: trip.destinations,
    days: trip.days.map((d) => ({ date: epochToDateOnly(d.date), city: d.city, title: d.title })),
    items: trip.items.map((it) => ({
      kind: it.kind,
      label: it.label,
      notes: it.notes,
      dayIndex: it.dayId !== undefined ? (dayIndex.get(it.dayId) ?? null) : null,
      time: it.time ?? null,
      owner: it.owner ?? null,
      isDone: it.isDone,
      sortOrder: it.sortOrder,
      reminderDate: it.reminderDate !== undefined ? epochToIso(it.reminderDate) : null,
      place: it.place
        ? { name: it.place.name, query: it.place.query,
            latitude: it.place.latitude ?? null, longitude: it.place.longitude ?? null }
        : null,
    })),
  };
  return JSON.stringify(dto, null, 2);
}

/** Parse a canonical export into a fresh-id Trip (tolerates absent or null optionals). */
export function importJSON(text: string): Trip {
  const dto = JSON.parse(text) as TripDTO;
  const now = Math.floor(Date.now() / 1000);
  const days: TripDay[] = (dto.days ?? []).map((d) => ({
    id: crypto.randomUUID(), date: dateOnlyToEpoch(d.date), city: d.city, title: d.title, modifiedAt: now,
  }));
  const items: ChecklistItem[] = (dto.items ?? []).map((i) => {
    const di = i.dayIndex;
    const dayId = di != null && di >= 0 && di < days.length ? days[di].id : undefined;
    const place: Place | undefined = i.place
      ? { name: i.place.name, query: i.place.query,
          latitude: i.place.latitude ?? undefined, longitude: i.place.longitude ?? undefined }
      : undefined;
    return {
      id: crypto.randomUUID(),
      kind: i.kind as ItemKind,
      label: i.label,
      notes: i.notes ?? "",
      dayId,
      time: i.time ?? undefined,
      owner: i.owner ?? undefined,
      isDone: i.isDone ?? false,
      sortOrder: i.sortOrder ?? 0,
      reminderDate: i.reminderDate != null ? isoToEpoch(i.reminderDate) : undefined,
      place,
      modifiedAt: now,
    };
  });
  return {
    id: crypto.randomUUID(),
    name: dto.name,
    startDate: dateOnlyToEpoch(dto.startDate),
    endDate: dateOnlyToEpoch(dto.endDate),
    destinations: dto.destinations ?? [],
    days, items, schemaVersion: 1, modifiedAt: now,
  };
}

// ── CSV (flat item-level, UTF-8 BOM) ───────────────────────────
const CSV_HEADER = "kind,label,notes,day_date,time,owner,is_done,place_name,place_query";

export function exportCSV(trip: Trip): string {
  const dayDate = new Map(trip.days.map((d) => [d.id, epochToDateOnly(d.date)]));
  const lines = [CSV_HEADER];
  for (const it of trip.items) {
    const cols = [
      it.kind, it.label, it.notes,
      it.dayId !== undefined ? (dayDate.get(it.dayId) ?? "") : "",
      it.time ?? "", it.owner ?? "", it.isDone ? "true" : "false",
      it.place?.name ?? "", it.place?.query ?? "",
    ];
    lines.push(cols.map(csvField).join(","));
  }
  return `﻿${lines.join("\n")}\n`;
}

/** Append CSV rows as items to a copy of `trip`, matching/creating a day by date. */
export function importCSVItems(csv: string, trip: Trip): Trip {
  const body = csv.startsWith("﻿") ? csv.slice(1) : csv;
  const rows = parseCSV(body);
  if (rows.length <= 1) return trip;
  const now = Math.floor(Date.now() / 1000);
  const byDate = new Map(trip.days.map((d) => [epochToDateOnly(d.date), d.id]));
  const days = [...trip.days];
  const items = [...trip.items];
  for (const row of rows.slice(1)) {
    if (row.length < 9) continue;
    let dayId: string | undefined;
    const d = row[3];
    if (d) {
      const existing = byDate.get(d);
      if (existing) dayId = existing;
      else {
        const id = crypto.randomUUID();
        days.push({ id, date: dateOnlyToEpoch(d), city: "", title: "", modifiedAt: now });
        byDate.set(d, id);
        dayId = id;
      }
    }
    const place: Place | undefined = row[7] ? { name: row[7], query: row[8] } : undefined;
    items.push({
      id: crypto.randomUUID(),
      kind: row[0] as ItemKind,
      label: row[1],
      notes: row[2],
      dayId,
      time: row[4] || undefined,
      owner: row[5] || undefined,
      isDone: row[6] === "true",
      sortOrder: items.length,
      place,
      modifiedAt: now,
    });
  }
  return { ...trip, days, items };
}

// RFC-4180-ish: quote fields containing comma/quote/newline; double inner quotes.
function csvField(s: string): string {
  if (!/[",\n]/.test(s)) return s;
  return `"${s.replace(/"/g, '""')}"`;
}

function parseCSV(text: string): string[][] {
  const rows: string[][] = [];
  let field = "";
  let row: string[] = [];
  let inQuotes = false;
  for (let i = 0; i < text.length; i++) {
    const c = text[i];
    if (inQuotes) {
      if (c === '"') {
        if (text[i + 1] === '"') { field += '"'; i++; }
        else inQuotes = false;
      } else field += c;
    } else if (c === '"') inQuotes = true;
    else if (c === ",") { row.push(field); field = ""; }
    else if (c === "\n") { row.push(field); rows.push(row); field = ""; row = []; }
    else if (c === "\r") { /* skip */ }
    else field += c;
  }
  if (field !== "" || row.length > 0) { row.push(field); rows.push(row); }
  return rows;
}

File: WanderIQKit/Sources/WanderIQKit/Export/TripExportCodec.swift
import Foundation

/// Canonical cross-platform trip import/export (format v1). Shared with the TS
/// codec (sub-project 6b) via Fixtures/trip-export-sample.json.
public enum TripExportCodec {

    // MARK: Wire DTOs (the canonical JSON shape)

    struct PlaceDTO: Codable { var name: String; var query: String
        var latitude: Double?; var longitude: Double? }
    struct DayDTO: Codable { var date: String; var city: String; var title: String }
    struct ItemDTO: Codable {
        var kind: String; var label: String; var notes: String
        var dayIndex: Int?; var time: String?; var owner: String?
        var isDone: Bool; var sortOrder: Int; var reminderDate: String?; var place: PlaceDTO?
    }
    struct TripDTO: Codable {
        var schemaVersion: Int; var name: String
        var startDate: String; var endDate: String; var destinations: [String]
        var days: [DayDTO]; var items: [ItemDTO]
    }

    private static let dayFmt: DateFormatter = {
        let f = DateFormatter(); f.calendar = Calendar(identifier: .iso8601)
        f.locale = Locale(identifier: "en_US_POSIX"); f.timeZone = TimeZone(identifier: "UTC")
        f.dateFormat = "yyyy-MM-dd"; return f
    }()
    private static let iso = ISO8601DateFormatter()

    // MARK: Export

    public static func exportJSON(_ trip: Trip) throws -> Data {
        let dayID = Dictionary(uniqueKeysWithValues: trip.days.enumerated().map { ($1.id, $0) })
        let dto = TripDTO(
            schemaVersion: 1, name: trip.name,
            startDate: dayFmt.string(from: trip.startDate),
            endDate: dayFmt.string(from: trip.endDate),
            destinations: trip.destinations,
            days: trip.days.map { DayDTO(date: dayFmt.string(from: $0.date), city: $0.city, title: $0.title) },
            items: trip.items.map { item in
                ItemDTO(kind: item.kind.rawValue, label: item.label, notes: item.notes,
                        dayIndex: item.dayID.flatMap { dayID[$0] }, time: item.time, owner: item.owner,
                        isDone: item.isDone, sortOrder: item.sortOrder,
                        reminderDate: item.reminderDate.map { iso.string(from: $0) },
                        place: item.place.map { PlaceDTO(name: $0.name, query: $0.query,
                                                         latitude: $0.latitude, longitude: $0.longitude) })
            })
        let enc = JSONEncoder(); enc.outputFormatting = [.prettyPrinted, .sortedKeys]
        return try enc.encode(dto)
    }

    // MARK: Import (always a fresh-id trip)

    public static func importJSON(_ data: Data) throws -> Trip {
        let dto = try JSONDecoder().decode(TripDTO.self, from: data)
        let now = Date()
        let days = dto.days.map { d in
            TripDay(date: dayFmt.date(from: d.date) ?? Date(timeIntervalSince1970: 0),
                    city: d.city, title: d.title, modifiedAt: now)
        }
        let items = dto.items.map { i -> ChecklistItem in
            let dayID: UUID? = i.dayIndex.flatMap { days.indices.contains($0) ? days[$0].id : nil }
            let place = i.place.map { Place(name: $0.name, query: $0.query,
                                            latitude: $0.latitude, longitude: $0.longitude) }
            return ChecklistItem(
                kind: ItemKind(rawValue: i.kind) ?? .prep, label: i.label, notes: i.notes,
                dayID: dayID, time: i.time, owner: i.owner, isDone: i.isDone, sortOrder: i.sortOrder,
                reminderDate: i.reminderDate.flatMap { iso.date(from: $0) }, place: place, modifiedAt: now)
        }
        return Trip(name: dto.name,
                    startDate: dayFmt.date(from: dto.startDate) ?? Date(timeIntervalSince1970: 0),
                    endDate: dayFmt.date(from: dto.endDate) ?? Date(timeIntervalSince1970: 0),
                    destinations: dto.destinations, days: days, items: items, modifiedAt: now)
    }

    // MARK: CSV (flat item-level, UTF-8 BOM)

    static let csvHeader = "kind,label,notes,day_date,time,owner,is_done,place_name,place_query"

    public static func exportCSV(_ trip: Trip) -> String {
        let dayDate = Dictionary(uniqueKeysWithValues: trip.days.map { ($0.id, dayFmt.string(from: $0.date)) })
        var lines = [csvHeader]
        for it in trip.items {
            let cols = [it.kind.rawValue, it.label, it.notes,
                        it.dayID.flatMap { dayDate[$0] } ?? "", it.time ?? "", it.owner ?? "",
                        it.isDone ? "true" : "false", it.place?.name ?? "", it.place?.query ?? ""]
            lines.append(cols.map(csvField).joined(separator: ","))
        }
        return "\u{FEFF}" + lines.joined(separator: "\n") + "\n"
    }

    /// Append CSV rows as items to `trip`, matching/creating a day by date.
    public static func importCSVItems(_ csv: String, into trip: inout Trip) {
        let body = csv.hasPrefix("\u{FEFF}") ? String(csv.dropFirst()) : csv
        let rows = parseCSV(body)
        guard rows.count > 1 else { return }
        let now = Date()
        var byDate = Dictionary(uniqueKeysWithValues: trip.days.map { (dayFmt.string(from: $0.date), $0.id) })
        for row in rows.dropFirst() where row.count >= 9 {
            var dayID: UUID?
            let d = row[3]
            if !d.isEmpty {
                if let existing = byDate[d] { dayID = existing }
                else if let date = dayFmt.date(from: d) {
                    let day = TripDay(date: date, city: "", title: "", modifiedAt: now)
                    trip.days.append(day); byDate[d] = day.id; dayID = day.id
                }
            }
            let place = row[7].isEmpty ? nil : Place(name: row[7], query: row[8])
            trip.items.append(ChecklistItem(
                kind: ItemKind(rawValue: row[0]) ?? .prep, label: row[1], notes: row[2],
                dayID: dayID, time: row[4].isEmpty ? nil : row[4],
                owner: row[5].isEmpty ? nil : row[5], isDone: row[6] == "true",
                sortOrder: trip.items.count, place: place, modifiedAt: now))
        }
    }

    // RFC-4180-ish: quote fields containing comma/quote/newline; double inner quotes.
    private static func csvField(_ s: String) -> String {
        guard s.contains(",") || s.contains("\"") || s.contains("\n") else { return s }
        return "\"" + s.replacingOccurrences(of: "\"", with: "\"\"") + "\""
    }

    private static func parseCSV(_ text: String) -> [[String]] {
        var rows: [[String]] = []; var field = ""; var row: [String] = []
        var inQuotes = false
        let chars = Array(text)
        var i = 0
        while i < chars.count {
            let c = chars[i]
            if inQuotes {
                if c == "\"" {
                    if i + 1 < chars.count, chars[i + 1] == "\"" { field.append("\""); i += 1 }
                    else { inQuotes = false }
                } else { field.append(c) }
            } else {
                switch c {
                case "\"": inQuotes = true
                case ",": row.append(field); field = ""
                case "\n": row.append(field); rows.append(row); field = ""; row = []
                case "\r": break
                default: field.append(c)
                }
            }
            i += 1
        }
        if !field.isEmpty || !row.isEmpty { row.append(field); rows.append(row) }
        return rows
    }
}
